### PR TITLE
Fix `tx/submit`

### LINKED ISSUE #29: Improve compatibillity with CardanoSharp-Wallet
[CardanoSharp-Wallet](https://github.com/CardanoSharp/cardanosharp-wallet) and blockfrost-dotnet combined will be a great tool to develop on the cardano blockchain. However, there's quite some room for improvements.

I wanted to create a Transaction that I previously serialized with CardanoSharp-Wallet. However, the data types of the two libraries do not match.

```C#
...
// Create serialized transaction using CardanoSharp.Wallet.Models.Transactions.Transaction
var signedTx = transaction.Serialize();

// Use blockfrost-dotnet to submit the transaction
transactionService.SubmitAsync(signedTx); // doesn't work because signedTxSerialized is byte[] but a string is expected
```

What also seems to be wrong is the way the data is transmitted. In `ABlockfrostService.SendPostRequestAsync()`, which is used for the above call to `transactionService.SubmitAsync(signedTx)`, the following code is present:

```C#
// line 345, pretty sure the StringContent doesn't work for binary data.
request_.Content = new StringContent(content, System.Text.Encoding.UTF8, "application/cbor");
```

### LINKED ISSUE #31: Fix tx/submit
Couple of issues here: 

- [x] #32
- [x] #33
- [x] #34
- [x] #35
- [x] #36

_Originally posted by @tweakch in https://github.com/blockfrost/blockfrost-dotnet/issues/30#issuecomment-903104217_

### LINKED ISSUE #32: Add predefined test vectors known to work on `cardano-cli`
- [ ] VectorId `00` 0 Ada tx
- [x] VectorId `01` 1 ada tx 
- [ ] VectorId `02` multi witness tx
- [ ] VectorId `03` metadata tx

### LINKED ISSUE #33: Create sender and receiver wallets for tests

### LINKED ISSUE #34: Submit transaction via stream
## Description

As described in #35 the Stream could be analyzed prior to sending the transaction. Invalid Streams should be rejected.

#### First byte

The first byte ***appears to (not verified through specific reference)*** always be a [major type 4](https://www.rfc-editor.org/rfc/rfc8949.html#name-major-types) followed by the number of elements N.

##### Example

An array (4 = `0b100`) of length 3 = `0b0011` we be represented as `0b100_0011` resp. `0x83` 

##### Verification

Peek at the first byte and verify that it is a valid Type 4. 

```cs
int type0 = 0x83 >> 4; // strip last four bits
bool valid = type0 == 4; // verify it is an array (4)
```

#### Second byte

The second byte ***apears to*** always be a map [major tyoe 5](https://www.rfc-editor.org/rfc/rfc8949.html#name-major-types) of the same length as `bytes[0]`. 

##### Example

An map (5 = `0b1001`) of length 3 = `0b0011` we be represented as `0b1001_0011` resp. `0xA3` 

##### Verfication

```cs
int type1 = 0xA3 >> 4;
int len0 = 0x83 & 0xf0;
int len1 = 0xA3 & 0xf0;
bool valid = (type1 == 5) && (len0 == len1);
```

[CBOR specification for reference](https://www.rfc-editor.org/rfc/rfc8949.html#name-specification-of-the-cbor-e)

### LINKED ISSUE #35: Submit transaction via string
## Description 

The `SubmitAsync(String)` method of the transaction service should inform the user that it expects the serialized and signed `cborHex` property of a transaction signed with `cardano-cli`. 

```json
{
  "type": "TxBodyMary",
  "description": "",
  "cborHex": "83a3008182582098035740ab68cad12cb4d8281d10ce1112ef0933dc84920b8937c3e80d78d12000018282581d60d0c43926a989c88d5049e61bdebf2a887aca10fa284b9067373ea28f0082581d603b75186909c120a97f6f0ee6822701f075e5136f3b9a08604a63dce700020080f6"
}
```

Meaning only hexadecimal characters are allowed: `/[0-9a-f]+/i`

#### Sample CBOR

```sh
83a3008182582098035740ab68cad12cb4d8281d10ce1112ef0933dc84920b8937c3e80d78d12000018282581d60d0c43926a989c88d5049e61bdebf2a887aca10fa284b9067373ea28f0082581d603b75186909c120a97f6f0ee6822701f075e5136f3b9a08604a63dce700020080f6
```

##### The above CBOR hex analyzed

```sh
83                                      # array(3)
   A3                                   # map(3)
      00                                # unsigned(0)
      81                                # array(1)
         82                             # array(2)
            58 20                       # bytes(32)
               98035740AB68CAD12CB4D8281D10CE1112EF0933DC84920B8937C3E80D78D120 # "\x98\x03W@\xABh\xCA\xD1,\xB4\xD8(\x1D\x10\xCE\x11\x12\xEF\t3\xDC\x84\x92\v\x897\xC3\xE8\rx\xD1 "
            00                          # unsigned(0)
      01                                # unsigned(1)
      82                                # array(2)
         82                             # array(2)
            58 1D                       # bytes(29)
               60D0C43926A989C88D5049E61BDEBF2A887ACA10FA284B9067373EA28F # "`\xD0\xC49&\xA9\x89\xC8\x8DPI\xE6\e\xDE\xBF*\x88z\xCA\x10\xFA(K\x90g7>\xA2\x8F"
            00                          # unsigned(0)
         82                             # array(2)
            58 1D                       # bytes(29)
               603B75186909C120A97F6F0EE6822701F075E5136F3B9A08604A63DCE7 # "`;u\x18i\t\xC1 \xA9\x7Fo\x0E\xE6\x82'\x01\xF0u\xE5\x13o;\x9A\b`Jc\xDC\xE7"
            00                          # unsigned(0)
      02                                # unsigned(2)
      00                                # unsigned(0)
   80                                   # array(0)
   F6                                   # primitive(22)
```

### Possible Improvements

We could also allow the actual CBOR (unhexed) and try to perform the hexdump within the Method to improve usability:

```sh
[{0: [[h'98035740AB68CAD12CB4D8281D10CE1112EF0933DC84920B8937C3E80D78D120', 0]], 1: [[h'60D0C43926A989C88D5049E61BDEBF2A887ACA10FA284B9067373EA28F', 0], [h'603B75186909C120A97F6F0EE6822701F075E5136F3B9A08604A63DCE7', 0]], 2: 0}, [], null]
```

### LINKED ISSUE #36: Verify transaction before submitting
Right now we only support `string` and `Stream` as datatypes to submit transactions. This can cause some confusion on what format the data provided to these methods must have in order to be accepted by the API and actually perform a transaction. 

To improve the dev experience we can perform some basic verification on the transactions. 

### Documentation

[Documentation](https://docs.blockfrost.io/#tag/Cardano-Transactions/paths/~1tx~1submit/post)

### SubmitAsync(String)

- [x] Submit TestVector `01` as hex string
- [x] Submit TestVector `01` as JSON (signed `cardano-cli` transaction)
- [ ] ~~#41~~

### SubmitAsync(Hex)

- [x] Submit TestVector `01` as FileStream from hexdump
- [x] Submit TestVector `01` as FileStream from JSON (signed `cardano-cli` transaction)
- [ ] ~~#40~~

### PR DESCRIPTION
**Pull request by tweakch**
## Description

This will fix the SDKs `tx/submit` endpoint 

Fixes #29 (partially)
Tracker #31 

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

In `Blockfrost.Api.Tests\Services\TransactionServiceTest.cs`
- [x] `TestSubmitAsyncString`
- [x] `TestSubmitAsyncStream`
- [x] Reject invalid strings (not hex)
- [x] Reject invalid strings (not json)
- [x] Reject invalid strings not cbor cddl)
- [x] Reject invalid streams (byte[0] is not array `0x81`,`0x82`,... `0x8f`)
- [x] Reject invalid streams (byte[1] is not map  `0xa1`,`0xa1`,...`0xaf`)
- [x] Reject invalid streams (octet[1] != octet[3] (f.e. `0x81` followed by `0xa2`) 

### Vectors

See [Vectors.md](https://github.com/blockfrost/blockfrost-dotnet/blob/feat/tx-submit/test/Blockfrost.Api.Tests/dat/Vectors.md)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by tweakch** [comment]
Couple of issues here: 

- [ ] Add predefined test vectors known to work on `cardano-cli`
- [ ] Create sender and receiver wallets for tests
- [ ] Submit transaction via stream
- [ ] Submit transaction via string 
- [ ] Verify transaction submit

File: test/Blockfrost.Api.Tests/dat/01/create.sh [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+#
+# Script to create the test vector
+SCRIPT=$(readlink -f "$0")
+VDIR=$(dirname $SCRIPT)
+DDIR=$(dirname $VDIR)
+AMOUNT=5000
+MAGIC=1097911063
```
Comment (tweakch) [inline test/Blockfrost.Api.Tests/dat/01/create.sh]: @mmahut @eddex If your bash skills are better than mine (I'd be surprised if they are not) feel free to improve this.

Comment (tweakch) [reply test/Blockfrost.Api.Tests/dat/01/create.sh]: and the code below... I'd waste my time here improving this.

File: test/Blockfrost.Api.Tests/dat/01/create.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+cardano-cli transaction build-raw \
+            --tx-in $_txIn#0 \
+            --tx-out $(cat $DDIR/keys/payment1.addr)+$_change \
+            --tx-out $(cat $DDIR/keys/payment2.addr)+$_out \
+            --fee $_fee \
+            --out-file $VDIR/tx.draft
+
+cardano-cli transaction sign \
+            --tx-body-file $VDIR/tx.draft \
+            --signing-key-file $DDIR/keys/payment1.skey \
```
Comment (tweakch) [inline test/Blockfrost.Api.Tests/dat/01/create.sh]: I'm not really concerned about the skey in the repo for now... if someone messes with it, i'll create a new wallet and hide it then... what do you think?

Comment (eddex) [reply test/Blockfrost.Api.Tests/dat/01/create.sh]: Should be ok. If someone decides to "steal" the tADA on the test wallet we can just get some more form the faucet.
I'd leave it like that for the sake of simplicity.

**Review by tweakch — COMMENTED** [review]
Works! 

*tested on ubuntu20.04 bash with synced `cardano-node` and `cardano-cli 1.27`*

```sh
$ pwd
~/blockfrost/blockfrost-dotnet/test/Blockfrost.Api.Tests
$ ./dat/01/create.sh 
Creating Test Vector 01...
done.
$ dotnet test --filter TestSubmitCardanoCliTransactionAsyncString
```

image

File: NOTICE [hunk 3]
```diff
@@ -0,0 +1,13 @@
+Copyright 2021 FIVE BINARIES OÜ
```
Comment (tweakch) [inline NOTICE:1]: This has to be included in order to limit liability, warranty and the other thing... i always forget the word

File: src/Blockfrost.Api/Constants.cs [hunk 4]
```diff
@@ -1,11 +1,11 @@
-﻿using System.Runtime.Serialization;
+﻿// Copyright (c) 2021 FIVE BINARIES OÜ. blockfrost-dotnet is licensed under the Apache License Version 2.0. See LICENSE in the project root for license information.
```
Comment (tweakch) [inline src/Blockfrost.Api/Constants.cs]: I personally cringe when I see these but  including this is not a bad idea from a legal point of view,

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,279 @@
+# EditorConfig is awesome:http://EditorConfig.org
```
Comment (tweakch) [inline .editorconfig:1]: I took the one from dotnet/format and edited it a little...

**Review by tweakch — COMMENTED** [review]
I think we can merge this and release v0.0.5

File: src/Blockfrost.Api/Constants.cs (hunk 4 above)
Comment (eddex) [reply src/Blockfrost.Api/Constants.cs]: Is it not enough to have the license in the root of the repo?

Comment (tweakch) [reply src/Blockfrost.Api/Constants.cs]: As far as I know (not a lawyer) it is.
@mmahut really up to you if we keep this or not

Comment (mmahut) [reply src/Blockfrost.Api/Constants.cs]: I think it is enough to have it in the root of the repo.

**Review by mmahut — APPROVED** [review]
Looks good, great job!

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+
+dotnet_naming_symbols.all_members.applicable_kinds = *
+
+dotnet_naming_style.pascal_case_style.capitalization = pascal_case
+
+# Analyzer diagnostincs: Warn by default, configure below
+dotnet_analyzer_diagnostic.severity = warning
+
+# IDE0073: File header
+dotnet_diagnostic.IDE0073.severity = none
```
Comment (tweakch) [inline .editorconfig:145]: remove warning if source file header is missing

**Review by tweakch — COMMENTED** [review]
removed all the headers

**Pull request merged**